common/src/main/java/com/thoughtworks/go/server/domain/AgentInstances.java

```diff
@@ -74,174 +85,133 @@ public AgentInstance findAgent(String uuid) {
-            if (agentInstance.isDisabled()){
-                agentInstances.add(agentInstance);
-            }
+        synchronized (uuidToAgentInstanceMap) {
+            stream(this.spliterator(), false)
+                    .filter(agentInstance -> agentInstance.getStatus().isRegistered())
+                    .forEach(registeredInstances::add);
         }
-        return agentInstances;
-    }
 
-    public AgentInstances findEnabledAgents() {
-        AgentInstances agentInstances = new AgentInstances(agentStatusChangeListener);
-        for (AgentInstance agentInstance : currentInstances()) {
-            if (agentInstance.getStatus().isEnabled()) {
-                agentInstances.add(agentInstance);
-            }
-        }
-        return agentInstances;
+        return registeredInstances;
     }
 
+    @Override
     public Iterator<AgentInstance> iterator() {
         return currentInstances().iterator();
     }
 
     public boolean isEmpty() {
-        return agentInstances.isEmpty();
-    }
-
-    public AgentInstance findFirstByHostname(String hostname) {
-        for (AgentInstance agentInstance : currentInstances()) {
-            if (agentInstance.agentConfig().getHostname().equals(hostname)) {
-                return agentInstance;
-            }
-        }
-        return new NullAgentInstance("");
+        return uuidToAgentInstanceMap.isEmpty();
     }
 
     public Integer size() {
-        return agentInstances.size();
-
+        return uuidToAgentInstanceMap.size();
     }
 
-    public void refresh() {
-        for (AgentInstance instance : currentInstances()) {
-            instance.refresh();
-        }
-        for (AgentInstance agentInstance : agentsToRemove()) {
-            removeAgent(agentInstance.agentConfig().getUuid());
-        }
+    public Agent findRegisteredAgentOrNull(String uuid){
+        AgentInstance agentInstance = findAgent(uuid);
+        return agentInstance.isRegistered() ? agentInstance.getAgent() : null;
     }
 
-    private List<AgentInstance> agentsToRemove() {
-        List<AgentInstance> agentsToRemove = new ArrayList<>();
-        for (AgentInstance instance : this) {
-            instance.checkForRemoval(agentsToRemove);
-        }
-        return agentsToRemove;
-    }
-
-    private Collection<AgentInstance> currentInstances() {
-        return new TreeSet<>(agentInstances.values());
+    public void refresh() {
+        currentInstances().forEach(AgentInstance::refresh);
+        getRemovableAgents().forEach(agentInstance -> removeAgent(agentInstance.getAgent().getUuid()));
     }
 
-    public void sync(Agents agentsFromConfig) {
-        for (AgentConfig agentInConfig : agentsFromConfig) {
-            String uuid = agentInConfig.getUuid();
-            if (agentInstances.containsKey(uuid)) {
-                agentInstances.get(uuid).syncConfig(agentInConfig);
+    public void syncAgentInstancesFrom(Agents agentsFromDB) {
```

Comment:
Should this be private?

Reply:
`syncAgentInstancesFrom` is called from `AgentService.syncAgentInstanceCacheFromAgentsInDB()`

Reply:
Sorry, my bad.